.github/CONTRIBUTING.md

```diff
@@ -53,68 +57,86 @@ issue. Stale issues will be closed.
-The `go test -short` flag can also be used to skip slower tests.
+### Modifying the Code
 
-Examples (run from the repository root):
-- `go test -v ./connect` will run all tests in the connect package (see `./connect` folder)
-- `go test -v -run TestRetryJoin ./command/agent` will run all tests in the agent package (see `./command/agent` folder) with name substring `TestRetryJoin`
+#### Code Formatting
 
-When a pull request is opened CI will run all tests and lint to verify the change.
+Go provides [tooling to apply consistent code formatting](https://golang.org/doc/effective_go#formatting).
+If you make any changes to the code, run `gofmt -s -w` to automatically format the code according to Go standards.
 
-## Go Module Dependencies
+#### Updating Go Module Dependencies
 
 If a dependency is added or change, run `go mod tidy` to update `go.mod` and `go.sum`.
 
-## Developer Documentation
+#### Developer Documentation
 
-Documentation about the Consul code base is under [./docs],
+Developer-focused documentation about the Consul code base is under [./docs],
 and godoc package document can be read at [pkg.go.dev/github.com/hashicorp/consul].
 
 [./docs]: ../docs/README.md
 [pkg.go.dev/github.com/hashicorp/consul]: https://pkg.go.dev/github.com/hashicorp/consul
 
-### Checklists
+### Testing
+
+During development, it may be more convenient to check your work-in-progress by running only the tests which you expect to be affected by your changes, as the full test suite can take several minutes to execute. [Go's built-in test tool](https://golang.org/pkg/cmd/go/internal/test/) allows specifying a list of packages to test and the `-run` option to only include test names matching a regular expression.
+The `go test -short` flag can also be used to skip slower tests.
+
+Examples (run from the repository root):
+- `go test -v ./connect` will run all tests in the connect package (see `./connect` folder)
+- `go test -v -run TestRetryJoin ./command/agent` will run all tests in the agent package (see `./command/agent` folder) with name substring `TestRetryJoin`
 
-Some common changes that many PRs require such as adding config fields, are
-documented through checklists.
+When a pull request is opened CI will run all tests and lint to verify the change.
 
-Please check in [docs/](../docs/) for any `checklist-*.md` files that might help
-with your change.
+### Submitting a Pull Request
+
+Before writing any code, we recommend:
+- Create a Github issue if none already exists for the code change you'd like to make.
+- Write a comment on the Github issue indicating you're interested in contributing so
+maintainers can provide their perspective if needed.
+
+When you're ready to submit a pull request:
```

Comment:
Not a blocker, but somewhere around here we might want to encourage contributors to keep their PRs small. Open a PR early with some questions to get feedback, to prevent them doing a lot of work that might be wasted.

Reply:
It's a good idea. I'll see if I can address in this PR.

Reply:
If I don't address as a part of this, I'm not sure when it will happen :) easy for that suggestion to get lost/forgotten.

Reply:
Going to add the below. Let me know what you think.

```
Keep your pull requests (PRs) small and open them early so you can get feedback on
approach from maintainers before investing your time in larger changes. For example,
see how [applying URL-decoding of resource names across the whole HTTP API](https://github.com/hashicorp/consul/issues/11258)
started with [iterating on the right approach for a few endpoints](https://github.com/hashicorp/consul/pull/11335)
before applying more broadly.
```

Reply:
Sounds great!